backend-rust/src/indexer.rs

```diff
@@ -335,96 +335,112 @@ impl Indexer for BlockPreProcessor {
+                    Ok(events)
+                };
 
-            let get_items = async move {
-                let items = client4
-                    .get_block_items(fbi.height)
-                    .await?
-                    .response
-                    .try_collect::<Vec<_>>()
-                    .await?;
-                Ok(items)
-            };
+                let get_items = async move {
+                    let items = client4
+                        .get_block_items(fbi.height)
+                        .await?
+                        .response
+                        .try_collect::<Vec<_>>()
+                        .await?;
+                    Ok(items)
+                };
 
-            let get_special_items = async move {
-                let items = client5
-                    .get_block_special_events(fbi.height)
-                    .await?
-                    .response
-                    .try_collect::<Vec<_>>()
-                    .await?;
-                Ok(items)
-            };
+                let get_special_items = async move {
+                    let items = client5
+                        .get_block_special_events(fbi.height)
+                        .await?
+                        .response
+                        .try_collect::<Vec<_>>()
+                        .await?;
+                    Ok(items)
+                };
+                let start_fetching = Instant::now();
+                let (block_info, chain_parameters, events, items, tokenomics_info, special_events) =
+                    try_join!(
+                        client1.get_block_info(fbi.height),
+                        client2.get_block_chain_parameters(fbi.height),
+                        get_events,
+                        get_items,
+                        client.get_tokenomics_info(fbi.height),
+                        get_special_items
+                    )?;
+                let get_total_staked_capital = async move {
+                    match tokenomics_info.response {
+                        RewardsOverview::V0 {
+                            ..
+                        } => {
+                            compute_total_stake_capital(
+                                &mut client6,
+                                BlockIdentifier::AbsoluteHeight(fbi.height),
+                            )
+                            .await
+                        }
+                        RewardsOverview::V1 {
+                            total_staked_capital,
+                            ..
+                        } => Ok(total_staked_capital),
+                    }
+                };
+                let get_block_certificates = async move {
+                    // Fetching the block certificates prior to P6 results in a NotFound gRPC error,
+                    // so we produce the empty type of certificates instead.
+                    if block_info.response.protocol_version < ProtocolVersion::P6 {
```

Comment:
It should really be done in the SDK

Reply:
This is not a bug, but an explicit choice for the gRPC API. Might not have been the right choice though.
Fixing it in the SDK, would mean we first have to fetch the block info to find the protocol version and then fetch it, which is not ideal either, so we do it here since we have the block_info anyway